apis/apps/v1alpha1/persistent_pod_state_types.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2022 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless persistent by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are persistent.  Any new fields you add must have json tags for the fields to be serialized.
+
+// PersistentPodStateSpec defines the desired state of PersistentPodState
+type PersistentPodStateSpec struct {
+	// TargetReference contains enough information to let you identify an workload for PersistentPodState
+	// Selector and TargetReference are mutually exclusive, TargetReference is priority to take effect
+	// current only support StatefulSet
+	TargetReference TargetReference `json:"targetRef"`
+
+	// Pod rebuilt topology required for node labels
+	// for examples: kubernetes.io/hostname, failure-domain.beta.kubernetes.io/zone
+	TopologyConstraint *TopologyTerm `json:"topologyConstraint,omitempty"`
+
+	// Pod rebuilt topology preferred for node labels, with xx weight
+	// for examples: kubernetes.io/hostname, failure-domain.beta.kubernetes.io/zone
+	TopologyPreference *TopologyPreference `json:"topologyPreference,omitempty"`
+}
+
+type TopologyPreference struct {
+	Weight     int32        `json:"weight"`
+	Preference TopologyTerm `json:"preference"`
+}
+type TopologyTerm struct {
+	// A list of node selector requirements by node's labels.
+	TopologyKeys []string `json:"topologyKeys"`
+}
+
+type PersistentPodStateStatus struct {
+	// observedGeneration is the most recent generation observed for this PersistentPodState. It corresponds to the
+	// PersistentPodState's generation, which is updated on mutation by the API Server.
+	ObservedGeneration int64 `json:"observedGeneration"`
+	// When the pod is ready, record some status information of the pod, such as: labels, annotations, topologies, etc.
+	// map[string]PodState -> map[Pod.Name]PodState
+	PodStates map[string]*PodState `json:"podStates,omitempty"`
+}
+
+type PodState struct {
+	// required topology key=value
+	TopologyConstraintState []TopologyStateTerm `json:"topologyConstraintState,omitempty"`
```

Comment:
Should we just record all required and preferred topologies in a single list? They should be splitted according `persistentPodState.spec` when webhook tries to inject them into the new pod.